Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp

```diff
@@ -0,0 +1,348 @@
+}
+
+void WebExtensionLocalization::loadRegionalLocalization(RefPtr<JSON::Object> regionalLocalization, RefPtr<JSON::Object> languageLocalization, RefPtr<JSON::Object> defaultLocalization, const String& bestLocale, const String& uniqueIdentifier)
+{
+    m_locale = WebCore::Locale::create(AtomString { bestLocale });
+    m_localeString = bestLocale;
+    m_uniqueIdentifier = uniqueIdentifier;
+
+    RefPtr localizationJSON = predefinedMessages();
+    localizationJSON = mergeJSON(localizationJSON, jsonWithLowercaseKeys(regionalLocalization));
+    localizationJSON = mergeJSON(localizationJSON, jsonWithLowercaseKeys(languageLocalization));
+    localizationJSON = mergeJSON(localizationJSON, jsonWithLowercaseKeys(defaultLocalization));
+
+    ASSERT(localizationJSON);
+
+    m_localizationJSON = localizationJSON;
+
+    return;
+}
+
+void WebExtensionLocalization::localizedJSONforJSON(RefPtr<JSON::Object> json)
+{
+    if (!json || !json->size())
+        return;
+
+    for (auto& key : json->keys()) {
+        auto valueType = json->getValue(key)->type();
+
+        if (valueType == JSON::Value::Type::String)
+            json->setString(key, localizedStringForString(json->getString(key)));
+        else if (valueType == JSON::Value::Type::Array)
+            localizedArrayForArray(json->getArray(key));
+        else if (valueType == JSON::Value::Type::Object)
+            localizedJSONforJSON(json->getObject(key));
+    }
+}
+
+String WebExtensionLocalization::localizedStringForKey(String key, Vector<String> placeholders)
+{
+    if (placeholders.size() > 9)
+        return emptyString();
+
+    if (!m_localizationJSON || !m_localizationJSON->size())
+        return emptyString();
+
+    if (key.contains("__MSG_"_s)) {
+        key = key.substring(6, key.length());
+        key = key.substring(0, key.length() - 2);
+    }
+
+    RefPtr localizationJSON = m_localizationJSON;
+    if (!localizationJSON)
+        return emptyString();
+
+    RefPtr stringJSON = localizationJSON->getObject(key.convertToASCIILowercase());
+    if (!stringJSON)
+        return emptyString();
+
+    auto localizedString = stringJSON->getString(messageKey);
+    if (!localizedString.length())
+        return emptyString();
+
+    RefPtr namedPlaceholders = jsonWithLowercaseKeys(stringJSON->getObject(placeholdersKey));
+
+    localizedString = stringByReplacingNamedPlaceholdersInString(localizedString, namedPlaceholders);
+    localizedString = stringByReplacingPositionalPlaceholdersInString(localizedString, placeholders);
+
+    localizedString = localizedString.impl()->replace("$$"_s, "$"_s);
+
+    return localizedString;
+}
+
+void WebExtensionLocalization::localizedArrayForArray(RefPtr<JSON::Array> json)
+{
+    if (!json)
+        return;
+
+    for (Ref value : *json) {
+        auto valueType = value->type();
+
+        if (valueType == JSON::Value::Type::String)
+            value->asString() = localizedStringForString(value->asString());
```

Comment:
@ItsJamie9494 This is incorrect and should be fixed.

Ideally both `localizedJSONforJSON()` and `localizedArrayForArray()` would return a modified copy, and not modify the input.

Reply:
to be clear, what's wrong here is modifying the original object instead of storing the localised string in a new object?

Reply:
Should be fixed in https://github.com/WebKit/WebKit/pull/37298

Reply:
That way the locale can change and the manifest substitutions could be resolved again in the new language. (We don't do a good job with that currently though.)